bin/test_hypothesis.py

```diff
@@ -0,0 +1,227 @@
+            [
+                sympy.sin,
+                sympy.cos,
+                sympy.tan,
+                sympy.asin,
+                sympy.acos,
+                sympy.atan,
+                sympy.sqrt,
+                sympy.powsimp,
+                sympy.expand_power_exp,
+                sympy.expand_power_base,
+                sympy.powdenest,
+                sympy.expand_log,
+                sympy.logcombine,
+                sympy.expand_func,
+                sympy.combsimp,
+                sympy.trigsimp,
+                sympy.expand_trig,
+                sympy.factorial,
+            ]
+        ),
+        x=expr,
+    )
+    def fn_expr(self, fn, x):
+        return fn(x)
+
+    @rule(target=expr, x=expr, y=expr, z=expr)
+    def subs_expr(self, x, y, z):
+        return x.subs(y, z)
+
+    @rule(target=expr, x=expr)
+    def simplify_expr(self, x):
+        return sympy.simplify(x)
+
+    @rule(target=expr, x=expr)
+    def cancel_expr(self, x):
+        return sympy.cancel(x)
+
+    @rule(target=expr, x=expr, force=st.booleans())
+    def expand_expr(self, x, force):
+        return sympy.expand(x, force=force)
+
+    @rule(target=expr, x=expr)
+    def factor_expr(self, x):
+        try:
+            return sympy.factor(x)
+        except ValueError:
+            reject()
+
+    @rule(target=expr, x=expr)
+    def apart_expr(self, x):
+        try:
+            return sympy.apart(x)
+        except (sympy.PolynomialError, sympy.NotAlgebraic, NotImplementedError):
+            reject()
+
+    @rule(target=expr, x=expr, y=expr)
+    def collect(self, x, y):
+        return sympy.collect(x, y)
+
+    @rule(x=expr)
+    def expr_doit(self, x):
+        x.doit()
+
+    @rule(x=expr)
+    def expr_evalf(self, x):
+        x.evalf()
+
+    @rule(
+        target=expr,
+        x=expr,
+        var=v,
+        f=c,
+        t=c,
+        op=st.sampled_from([sympy.Sum, sympy.Product]),
+    )
+    def combine_many(self, x, var, f, t, op):
+        return op(x, (var, f, t))
+
+    @rule(source=expr, data=st.data())
+    def substitution_commutes_with_simplify(self, source, data):
+        symbols = sorted(
+            [v for v in source.atoms() if isinstance(v, sympy.Symbol)],
+            key=lambda a: a.name,
+        )
+
+        if not symbols:
+            return
+
+        variables = data.draw(st.permutations(symbols), label="variables")
+        assignment = [sympy.Integer(data.draw(st.integers())) for _ in variables]
+
+        r1 = source.subs(dict(zip(variables, assignment)))
+
+        if isinstance(r1, sympy.numbers.NaN) or r1 != r1:
+            return
+
+        r2 = source
+        for a, v in zip(variables, assignment):
+            r2 = sympy.simplify(r2.subs({a: v}))
+
+        if isinstance(r2, sympy.numbers.NaN) or r2 != r2:
+            return
+
+        r1 = r1.doit()
+        r2 = r2.doit()
+
+        assert (r1.doit() == r2.doit()) or (abs(complex(r1) - complex(r2)) < EPS)
```

Comment:
```suggestion
        assert (r1 == r2) or (abs(complex(r1) - complex(r2)) < EPS)
```

Reply:
Unfortunately removing these `.doit()` calls leads to tests that hang forever (or at least half an hour) :sob:

Reply:
doit() is already called above, and it should be idempotent. So whatever causes this is a bug.

Reply:
Unfortunately not one that I could detect by adding a method; the following still passes.

```python
    @rule(source=expr)
    def doit_is_idempotent(self, source):
        done = source.doit()
        assert done == done.doit()
```

Reply:
Is it possible to interrupt the hang and see what causes it? Is it possible to run these tests in some kind of verbose mode that will show what is being tested before it hangs?

Reply:
Probably you should add timeout functionality since it's very easy to create expressions that will take a very long time to simplify / evaluate.

Reply:
Yes to both - you can set a timeout that will interrupt the test using the stdlib `signal` module on Linux, or the pypi `stopit` module if you don't mind a dependency.

Verbose mode can be enabled by setting `Verbosity.normal` to `.verbose` (or `.debug` to see every method call) in the `@settings(...)` decorator on the class.

Reply:
Can we add those as command line options?

I think having stopit as a dependency is fine since this isn't really for end users. We should have something that works on Windows/OSX though since many contributors don't use Linux.

On CI I think maybe we should use the timeout and it should be reported like a test failure with details to reproduce. I don't know how far this currently goes in creating complicated expressions that actually should take a long time though so we might find that the timeout is creating false positive test failures but we can evaluate that after trying it for a bit.

Seeing what the output looks like I guess that setting full verbosity would create an absurdly large log so that probably shouldn't happen by default or on Travis.

Reply:
> Can we add those as command line options?

`pytest bin/test_hypothesis.py --hypothesis-verbosity={quiet,normal,verbose,debug}` all work - Hypothesis ships with a simple pytest plugin 😄

I'll add a dependency on `stopit` too; since the interrupt raises an exception that should automatically be treated as a failure.